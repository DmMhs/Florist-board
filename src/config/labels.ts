const labels = {
  brand: 'FLORIST.UA',
  adress: 'MARKET SQUARE, LVIV, UA',
  postCode: '346346',
  phone: '+123-45-678',
  navigation: {
    home: 'HOME',
    shop: 'SHOP',
    gallery: 'GALLERY',
    contacts: 'CONTACTS',
    cart: 'CART',
    account: {
      main: 'MY ACCOUNT',
      menu: {
        signIn: 'SIGN IN',
        signUp: 'SIGN UP'
      }
    }
  }
};

export default labels;
