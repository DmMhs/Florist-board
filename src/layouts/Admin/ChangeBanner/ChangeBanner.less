.AddGalleryImage,
.ChangeDesktopBanner {
  .form-control {
    select {
      margin-left: 8px;
    }
  }
}
