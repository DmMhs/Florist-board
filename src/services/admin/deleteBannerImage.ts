import { homeImagesRef, storageRef } from '../../firebase';

const getKeyByValue = (object: { [key: string]: string }, value: string) => {
  return Object.keys(object).find(key => object[key] === value);
};

const map = { first: '1', second: '2' };

export const deleteBannerImage = (imageName: string, imageUrl: string) => {
  const confirm = window.confirm('Are you sure?');
  if (confirm) {
    return storageRef
      .child('banner')
      .child(imageName)
      .delete()
      .then(response => {
        return homeImagesRef
          .once('value')
          .then(snapshot => {
            const imageKey = getKeyByValue(snapshot!.val(), imageUrl);
            return homeImagesRef.child(imageKey as string).remove();
          })
          .catch(err => console.log(err));
      });
  }
};
