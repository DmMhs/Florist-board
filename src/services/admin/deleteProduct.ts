import { productsRef } from '../../firebase';
import { deleteProductImages } from './deleteProductImages';

export const deleteProduct = async (id: string, name: string) => {
  const confirm = window.confirm('Are you sure?');
  if (confirm) {
    await deleteProductImages(id, name);
    return productsRef
      .child(id)
      .remove()
      .catch(err => console.log(err));
  }
};
