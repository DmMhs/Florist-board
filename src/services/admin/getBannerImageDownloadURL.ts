import { storageRef } from '../../firebase';

export const getBannerImageDownloadURL = async (fileName: string) => {
  return storageRef
    .child('banner')
    .child(fileName)
    .getDownloadURL();
};
