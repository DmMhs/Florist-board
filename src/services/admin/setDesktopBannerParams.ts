import { database } from '../../firebase';

export const setDesktopBannerParams = (width: string, height: string) => {
  return database.ref('banner-params').set({
    desktopBannerWrapperHeight: height,
    desktopBannerWrapperWidth: width
  });
};
