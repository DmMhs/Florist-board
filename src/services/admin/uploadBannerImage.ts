import { storageRef } from '../../firebase';

export const uploadBannerImage = (file: File, fileName: string) => {
  return storageRef
    .child('banner')
    .child(fileName)
    .put(file)
    .catch(err => console.log(err));
};
