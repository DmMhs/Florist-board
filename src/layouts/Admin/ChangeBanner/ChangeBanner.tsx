import React, { Component } from 'react';

import { AppContext } from '../../../AppContext';
import { storageRef, homeImagesRef, database } from '../../../firebase';
import './ChangeBanner.less';
import { uploadBannerImage } from '../../../services/admin/uploadBannerImage';
import { getBannerImageDownloadURL } from '../../../services/admin/getBannerImageDownloadURL';

interface AddBannerImageProps {}

interface AddBannerImageState {
  images: File[];
  totalImagesNumber: number;
  desktopBannerWrapperWidth: string;
  desktopBannerWrapperHeight: string;
  bannerWidthUnits: string;
  bannerHeightUnits: string;
}

class AddBannerImage extends Component<
  AddBannerImageProps,
  AddBannerImageState
> {
  constructor(props: AddBannerImageProps) {
    super(props);
    this.state = {
      images: [],
      totalImagesNumber: 0,
      desktopBannerWrapperHeight: '',
      desktopBannerWrapperWidth: '',
      bannerWidthUnits: '%',
      bannerHeightUnits: 'px'
    };
  }

  public componentDidMount() {
    homeImagesRef.on('value', snapshot => {
      this.setState({
        totalImagesNumber: Object.keys(snapshot!.val()).length
      });
    });
  }

  private imgInputChangedHandler = (
    position: number,
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const updatedImages: File[] = [...this.state.images];
    updatedImages[position] = event.target.files as any;
    this.setState({
      images: updatedImages
    });
  };

  private bannerImageAddedHandler = (
    event: React.FormEvent<HTMLFormElement>
  ) => {
    event.preventDefault();
    const { images } = this.state;
    const {} = this.state;
    this.setState({
      images: []
    });
    images.map(async (image: File) => {
      const file = (image as any)[0];
      const formattedFileName = (image as any)[0].name;
      await uploadBannerImage(file, formattedFileName);
      getBannerImageDownloadURL(formattedFileName)
        .then(res => {
          return res;
        })
        .then(imgURL => {
          homeImagesRef.push(imgURL);
        });
    });
  };

  private bannerChangedHandler = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const {
      desktopBannerWrapperHeight,
      desktopBannerWrapperWidth,
      bannerWidthUnits,
      bannerHeightUnits
    } = this.state;
    const {} = this.state;
    const width =
      desktopBannerWrapperWidth === ''
        ? '0px'
        : desktopBannerWrapperWidth + bannerWidthUnits;
    const height =
      desktopBannerWrapperHeight === ''
        ? '0px'
        : desktopBannerWrapperHeight + bannerHeightUnits;
    database.ref('banner-params').set({
      desktopBannerWrapperHeight: height,
      desktopBannerWrapperWidth: width
    });
  };

  private widthInputChangedHandler = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    this.setState({
      desktopBannerWrapperWidth: event.target.value
    });
  };

  private widthSelectChangedHandler = (
    event: React.ChangeEvent<HTMLSelectElement>
  ) => {
    this.setState({
      bannerWidthUnits: event.target.value
    });
  };

  private heightInputChangedHandler = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    this.setState({
      desktopBannerWrapperHeight: event.target.value
    });
  };

  private heightSelectChangedHandler = (
    event: React.ChangeEvent<HTMLSelectElement>
  ) => {
    this.setState({
      bannerHeightUnits: event.target.value
    });
  };

  public render() {
    const context = this.context;
    const labels = context.state.labels;
    const lang = context.state.lang;

    const labelsRoot = labels[lang].pages.admin.changeBannerForm;
    const submitBtnLabel = labels[lang].pages.admin.submitBtn;

    return (
      <AppContext.Consumer>
        {value =>
          value && (
            <div>
              <form
                onSubmit={this.bannerImageAddedHandler}
                className="AddGalleryImage form"
              >
                <div className="form-control product-images">
                  <p className="gallery-info">
                    {labelsRoot.info.totalImagesNumber}{' '}
                    {this.state.totalImagesNumber}
                  </p>
                  <hr />
                  <label className="accent">{labelsRoot.add}</label>
                  <br />
                  <div className="input-wrapper">
                    <input
                      type="file"
                      onChange={this.imgInputChangedHandler.bind(this, 0)}
                    />
                  </div>
                </div>
                <button type="submit">{submitBtnLabel}</button>
              </form>

              <form
                className="ChangeDesktopBanner form"
                onSubmit={this.bannerChangedHandler}
              >
                <div className="form-control">
                  <label className="accent">
                    {labelsRoot.change.desktopWrapperWidth}
                  </label>
                  <br />
                  <input
                    type="number"
                    onChange={this.widthInputChangedHandler}
                  />
                  <select
                    onChange={this.widthSelectChangedHandler}
                    defaultValue={this.state.bannerWidthUnits}
                  >
                    <option value="px">px</option>
                    <option value="%">%</option>
                    <option value="rem">rem</option>
                    <option value="em">em</option>
                    <option value="vw">vw</option>
                  </select>
                </div>
                <div className="form-control">
                  <label className="accent">
                    {labelsRoot.change.desktopWrapperHeight}
                  </label>
                  <br />
                  <input
                    type="number"
                    onChange={this.heightInputChangedHandler}
                  />
                  <select
                    onChange={this.heightSelectChangedHandler}
                    defaultValue={this.state.bannerHeightUnits}
                  >
                    <option value="px">px</option>
                    <option value="%">%</option>
                    <option value="rem">rem</option>
                    <option value="em">em</option>
                    <option value="vw">vh</option>
                  </select>
                </div>
                <button type="submit">{submitBtnLabel}</button>
              </form>
            </div>
          )
        }
      </AppContext.Consumer>
    );
  }
}

AddBannerImage.contextType = AppContext;

export default AddBannerImage;
