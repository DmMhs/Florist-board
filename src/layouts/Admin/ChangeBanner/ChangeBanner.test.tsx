import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';

import ChangeBanner from './ChangeBanner';
import AppContextProvider, { AppContext } from '../../../AppContext';
import labels from '../../../config/labels';
import * as getBannerImageDownloadURLFunction from '../../../services/admin/getBannerImageDownloadURL';
import * as uploadBannerImageFunction from '../../../services/admin/uploadBannerImage';
import * as setDesktopBannerParamsFunction from '../../../services/admin/setDesktopBannerParams';

describe('ChangeBanner works as expected', () => {
  it('ChangeBanner component matches a snapshot', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <ChangeBanner />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
  it('submit event envokes helper functions', async () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <ChangeBanner />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('ChangeBanner').instance();
    instance.setState({
      images: ['one']
    });

    const p = Promise.resolve('success');
    const spyUpload = jest
      .spyOn(uploadBannerImageFunction, 'uploadBannerImage')
      .mockImplementation(() => p);
    const spyGetDownloadURL = jest
      .spyOn(getBannerImageDownloadURLFunction, 'getBannerImageDownloadURL')
      .mockImplementation(() => p);

    instance.forceUpdate();

    wrapper.find('.AddGalleryImage').simulate('submit');
    wrapper.update();
    await p;
    expect(instance.state).toEqual({
      images: [],
      totalImagesNumber: 0,
      desktopBannerWrapperHeight: '',
      desktopBannerWrapperWidth: '',
      bannerWidthUnits: '%',
      bannerHeightUnits: 'px'
    });
    expect(spyUpload).toHaveBeenCalled();
    expect(spyGetDownloadURL).toHaveBeenCalled();
  });

  it('Changing banner parameters envokes helper functions', async () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <ChangeBanner />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('ChangeBanner').instance();
    instance.setState({
      images: ['one']
    });

    const p = Promise.resolve('success');
    const spySetParams = jest
      .spyOn(setDesktopBannerParamsFunction, 'setDesktopBannerParams')
      .mockImplementation(() => p);
    instance.forceUpdate();

    wrapper.find('.ChangeDesktopBanner').simulate('submit');
    wrapper.update();
    await p;
    expect(spySetParams).toHaveBeenCalled();
  });
  it('Event handlers change the state', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <ChangeBanner />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('ChangeBanner').instance();
    instance.setState({
      images: ['one']
    });
    instance.forceUpdate();
    wrapper.find('.widthInput').simulate('change', { target: { value: '10' } });
    wrapper
      .find('.widthSelect')
      .simulate('change', { target: { value: 'px' } });
    wrapper
      .find('.heightInput')
      .simulate('change', { target: { value: '30' } });
    wrapper
      .find('.heightSelect')
      .simulate('change', { target: { value: '%' } });
    wrapper.update();
    expect(instance.state).toEqual({
      images: ['one'],
      totalImagesNumber: 0,
      desktopBannerWrapperHeight: '30',
      desktopBannerWrapperWidth: '10',
      bannerWidthUnits: 'px',
      bannerHeightUnits: '%'
    });
  });
});
