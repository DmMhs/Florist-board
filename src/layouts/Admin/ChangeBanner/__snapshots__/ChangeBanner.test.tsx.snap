// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeBanner works as expected ChangeBanner component matches a snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(AppContextProvider)>
      <Route>
        <AppContextProvider
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <ChangeBanner>
            <div>
              <form
                className="AddGalleryImage form"
                onSubmit={[Function]}
              >
                <div
                  className="form-control product-images"
                >
                  <p
                    className="gallery-info"
                  >
                    Number of banner images: 
                     
                    0
                  </p>
                  <hr />
                  <label
                    className="accent"
                  >
                    Add new banner image:
                  </label>
                  <br />
                  <div
                    className="input-wrapper"
                  >
                    <input
                      onChange={[Function]}
                      type="file"
                    />
                  </div>
                </div>
                <button
                  type="submit"
                >
                  SUBMIT
                </button>
              </form>
              <form
                className="ChangeDesktopBanner form"
                onSubmit={[Function]}
              >
                <div
                  className="form-control"
                >
                  <label
                    className="accent"
                  >
                    Change width of banner wrapper on desktop devices:
                  </label>
                  <br />
                  <input
                    className="widthInput"
                    onChange={[Function]}
                    type="number"
                  />
                  <select
                    className="widthSelect"
                    defaultValue="%"
                    onChange={[Function]}
                  >
                    <option
                      value="px"
                    >
                      px
                    </option>
                    <option
                      value="%"
                    >
                      %
                    </option>
                    <option
                      value="rem"
                    >
                      rem
                    </option>
                    <option
                      value="em"
                    >
                      em
                    </option>
                    <option
                      value="vw"
                    >
                      vw
                    </option>
                  </select>
                </div>
                <div
                  className="form-control"
                >
                  <label
                    className="accent"
                  >
                    Change height of banner wrapper on desktop devices:
                  </label>
                  <br />
                  <input
                    className="heightInput"
                    onChange={[Function]}
                    type="number"
                  />
                  <select
                    className="heightSelect"
                    defaultValue="px"
                    onChange={[Function]}
                  >
                    <option
                      value="px"
                    >
                      px
                    </option>
                    <option
                      value="%"
                    >
                      %
                    </option>
                    <option
                      value="rem"
                    >
                      rem
                    </option>
                    <option
                      value="em"
                    >
                      em
                    </option>
                    <option
                      value="vw"
                    >
                      vh
                    </option>
                  </select>
                </div>
                <button
                  type="submit"
                >
                  SUBMIT
                </button>
              </form>
            </div>
          </ChangeBanner>
        </AppContextProvider>
      </Route>
    </withRouter(AppContextProvider)>
  </Router>
</BrowserRouter>
`;
